### ISSUE TITLE
Parsing gc equations

### ISSUE THREAD

**Issue by andrcoding** [cgeo/cgeo#12517]
### Tell us your idea!

I am not quite sure if I use the app correctly. However, I often copy equations from the gc description into the waypoint calculation. In many gc the sign 'x' is used as '*' and causes issues at value calculation.

### What solution would you suggest?

Therefore, I would find it useful if there would be a function in the submenu, as a button, or somehow similar to convert 'x' into ' * ' so that equations like 3x5 are handled "correctly": 3*5.

**Comment by eddiemuc** [comment]
The new calculator already has a function "remove spaces". Should we rename this to e.g. "clean up", combine the space removal with replacement of x with *, and maybe in time there will be more things coming into this?

**Comment by ztNFny** [comment]
I thought the recognition of chars like /, x, .. was already implemented?

**Comment by SammysHP** [comment]
What if it is `… 36.3x9 … 02.1y4`? Hopefully `XxY` is not out there in the wild. ;)

**Comment by eddiemuc** [comment]
> I thought the recognition of chars like /, x, .. was already implemented?

Only in the new implementation, not the one in production.

Also, recognition of x as * was not implemented because of ambiguous expressions like 6x3 (without spaces) -> is x a variable or a multiplicator symbol?

If we exclude the single lowercase letter x from being used as a variable, then x can be interpreted as multiplicator unambiguous. Or  it is only interpreted as such if there are spaces e.g. `6 x 3`?

**Comment by ztNFny** [comment]
> Only in the new implementation

That's clear, I'm only talking about the new one.

> Also, recognition of x as * was not implemented because of ambiguous expressions like 6x3 (without spaces) -> is x a variable or a multiplicator symbol?

IIRC we agreed that the conversion of ambigious chars (x, en-dash, em-dash, ...) happens when recognizing the formula, then afterwards only calculation chars are used which allows for manual correction later on? Can't find the ticket now though.

**Comment by eddiemuc** [comment]
> IIRC we agreed that the conversion of ambigious chars (x, en-dash, em-dash, ...) happens when recognizing the formula, then afterwards only calculation chars are used which allows for manual correction later on? Can't find the ticket now though.

I remember that discussion too. But the "scan formula" function is currently implemented only rudimentary, and there's no "scan waypoints" function yet which recognizes formulas. Perhaps a new issue might be best to track this

**Comment by eddiemuc** [comment]
> IIRC we agreed that the conversion of ambigious chars (x, en-dash, em-dash, ...) happens when recognizing the formula

One more thing: Input recognition of em-dash, em-dash and some others into normal formula parser because their recognition is unambiguous. Only problems I remember were the x and the [] brackets (at the time there was discussion to use [] to specify ranges)

**Comment by ztNFny** [comment]
> But the "scan formula" function is currently implemented only rudimentary

ah, so that's only in the "scan" function. Then I'd propose to add a button "parse formula" with the same logic to the formula input dialog.

> and there's no "scan waypoints" function yet which recognizes formulas.

That's the most important missing feature in that area imho. Do we have a dedicated issue to track that or is it only stored in brain.exe?

> I put recognition of em-dash, em-dash and some others into normal formula parser because their recognition is unambiguous. Only problems I remember were the x and the [] brackets (at the time there was discussion to use [] to specify ranges)

Imho "x" is more commonly used as "*" than as variable. The recognition will never be 100% correct, so we should strive for 80% and for unambigious cases implement the more-likely-case - as long as the user can correct it afterwards everything is ok.

**Comment by eddiemuc** [comment]
> > But the "scan formula" function is currently implemented only rudimentary
> 
> ah, so that's only in the "scan" function. Then I'd propose to add a button "parse formula" with the same logic to the formula input dialog.

Interesting idea. As said before, there is already a button named "remove spaces" on waypoint formula field. Maybe we can rename this to something like "parse formula", add it also to the variables in the var tab and put all "cleanup" functions into it which come to our mind. Right now this would only be "remove all spaces" and "replace x with *", but I am sure there will be more.

> > and there's no "scan waypoints" function yet which recognizes formulas.
> 
> That's the most important missing feature in that area imho. Do we have a dedicated issue to track that or is it only stored in brain.exe?

That would be issue #12383 

 
> > I put recognition of em-dash, em-dash and some others into normal formula parser because their recognition is unambiguous. Only problems I remember were the x and the [] brackets (at the time there was discussion to use [] to specify ranges)
> 
> Imho "x" is more commonly used as "*" than as variable. The recognition will never be 100% correct, so we should strive for 80% and for unambigious cases implement the more-likely-case - as long as the user can correct it afterwards everything is ok.

So does this now mean that a `x` should never be considered a var (except in constructs like `$x` or `${x}` or maybe `x x 6` (first x is probably a var and second is not - not sure how hard this will be to implement though)) and always as `*`?

**Comment by ztNFny** [comment]
> So does this now mean that a x should never be considered a var (except in constructs like $x or ${x} or maybe x x 6 (first x is probably a var and second is not - not sure how hard this will be to implement though)) and always as *?

In an actual formula a letter should always be a variable, only *+-/ should be considered expression chars.
But in the "scan" function some conversion should happen (en-dash -> -, x -> *, [] -> (), ...) and there I suggest to treat "x" as "*"

> Right now this would only be "remove all spaces" and "replace x with *", but I am sure there will be more.

I'd proposed to use the same "recognition" logic that the scan listing function uses. So that you get the same result whether you use "scan listing" or copy & paste the formula from listing, then use the "scan / parse" button there.

**Comment by eddiemuc** [comment]
> > Right now this would only be "remove all spaces" and "replace x with *", but I am sure there will be more.
> 
> I'd proposed to use the same "recognition" logic that the scan listing function uses. So that you get the same result whether you use "scan listing" or copy & paste the formula from listing, then use the "scan / parse" button there.

The current scan function is regex-based (which is probably not a good idea anyway). It is currently not doing very much with the found texts:
* it replaces "x" by "*", and
* it replaces horizontal whitespace with normal space

See class `FormulaUtils`

![image](https://user-images.githubusercontent.com/6909759/149514618-4351a8fe-5c97-45e7-87b0-9beacae70532.png)

**Comment by ztNFny** [comment]
ah ok, I thought those conversion rules we had discussed in that other issue was already implemented

**Comment by eddiemuc** [comment]
> ah ok, I thought those conversion rules we had discussed in that other issue was already implemented

Currently it is implemented like this:
* "x" -> "*" rule is implemented in "scan formula" only (as described above)
* "[]" -> "()" is not implemented at all
* all other rules are unambigious (e.g.  – – — for minus or • for multiplication) and thus are implemented directly in the formula parser. 

I hope I am not forgetting anything, but at least this should be the state to the best of my knowledge.
BTW: I found the issue again where we discussed all this, it is #12316

**Comment by ztNFny** [comment]
The idea I had was that there's one "cleanup" function that does all those replacements (both x and —), then no "intelligence" at all in the formula parser. Like that anything other than a mathematical operator could "stupidly" be parsed as a variable in the formula parser, let's say the "x" or some chinese character. 

The danger of doing any intelligence in the formula parser is that it interprets things wrong and the user has no chance to see that. Granted that for "– – —" -> minus the likelyhood of there being a wrong interpretation is near-zero, but I'm sure over time there will be ambigious cases.

**Comment by eddiemuc** [comment]
@ztNFny I fully agree with your statement, which is why I only put things in the parser which are imho unambiguous (and not the "x" parsing, because this is not)

**Comment by Lineflyer** [comment]
@andrcoding 
Also what is discussed here refers to the new version currently being implemented as mentioned in the other issue.

Once again:
It would be really good to get user feedback on the complete new implementation.
We do now have a "cache wide" variable tab and a reworked waypoint calculator making use of it.

It will take some time until the new version is live and maybe the new calculator will then still be hidden (only available in debug mode). But if you want to try it out and provide early feedback to us (which would be very helpful) you can think about upgrading to our nightly versions (see https://www.cgeo.org/development) where it is already usable in its current development state.

**Comment by andrcoding** [comment]
Yes, I have seen it. The new version has obviously some improvements I really like. My time in the upcoming time is limited for gc but I will give as many feedback as possible. For the next gc walk (not before Feb) I will check out the nightly version. Thanks for your work!

**Comment by murggel** [comment]
Fixed with PR #16914
